File: July 2024/24-07-2024/Check for BST.cs
//{ Driver Code Starts
// Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

class Node {
    public int data;
    public Node left;
    public Node right;

    public Node(int key) {
        this.data = key;
        this.left = null;
        this.right = null;
    }
}

namespace DriverCode {
class GFG {
    // Function to Build Tree
    public Node buildTree(string str) {
        // Corner Case
        if (str.Length == 0 || str[0] == 'N') return null;

        // Creating vector of strings from input
        // string after spliting by space
        var ip = str.Split(' ');

        Node root = new Node(int.Parse(ip[0]));

        // Push the root to the queue
        Queue<Node> queue = new Queue<Node>();
        queue.Enqueue(root);

        // Starting from the second element
        int i = 1;
        while (queue.Count != 0 && i < ip.Length) {

            // Get and remove the front of the queue
            Node currNode = queue.Peek();
            queue.Dequeue();

            // Get the current node's value from the string
            string currVal = ip[i];

            // If the left child is not null
            if (currVal != "N") {

                // Create the left child for the current node
                currNode.left = new Node(int.Parse(currVal));

                // Push it to the queue
                queue.Enqueue(currNode.left);
            }

            // For the right child
            i++;
            if (i >= ip.Length) break;
            currVal = ip[i];

            // If the right child is not null
            if (currVal != "N") {

                // Create the right child for the current node
                currNode.right = new Node(int.Parse(currVal));

                // Push it to the queue
                queue.Enqueue(currNode.right);
            }
            i++;
        }

        return root;
    }

    void preorder(Node root) {
        if (root == null) {
            return;
        }
        Console.Write(root.data + " ");
        preorder(root.left);
        preorder(root.right);
    }

    void inorder(Node root) {
        if (root == null) {
            return;
        }

        inorder(root.left);
        Console.Write(root.data + " ");
        inorder(root.right);
        return;
    }

    void postorder(Node root) {
        if (root == null) {
            return;
        }

        postorder(root.left);
        postorder(root.right);
        Console.Write(root.data + " ");
    }

    static void Main(string[] args) {
        int testcases; // Taking testcase as input
        testcases = Convert.ToInt32(Console.ReadLine());
        while (testcases-- > 0) // Looping through all testcases
        {
            var gfg = new GFG();
            var str = Console.ReadLine().Trim();
            var root = gfg.buildTree(str);
            Solution obj = new Solution();
            var res = obj.isBST(root);

            Console.WriteLine(res ? "true" : "false");
        }
    }
}
}
// } Driver Code Ends


// User function Template for C#

class Solution {
    // Complete this function
    // Function to check whether a Binary Tree is BST or not.
    
    //Your Code Start
    private Node prev;
    private bool InTrav(Node root) {
        if (root == null) {
            return true;
        }

        if (!InTrav(root.left)) {
            return false;
        }

        if (prev != null && root.data <= prev.data) {
            return false;
        }

        prev = root;

        return InTrav(root.right);
    }

    public bool isBST(Node root) {
        prev = null;
        return InTrav(root);
    }
    //Your Code End
    
}

File: July 2024/24-07-2024/Check for BST.js
//{ Driver Code Starts
// Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

function newNode(root, data) {
    if (root === null)
        root = new Node(data);
    else if (data < root.data)
        root.left = newNode(root.left, data);
    else
        root.right = newNode(root.right, data);
    return root;
}

function buildTree(str) {
    // Corner Case
    if (str.length === 0 || str[0] === "N") return null;

    // Create the root of the tree
    let root = new Node(parseInt(str[0]));

    // Push the root to the queue
    let queue = [];
    let start = 0;
    queue.push(root);

    // Starting from the second element
    let i = 1;
    while (queue.length !== start && i < str.length) {

        // Get and remove the front of the queue
        let currNode = queue[start];
        start++;

        // Get the current node's value from the string
        let currVal = str[i];

        // If the left child is not null
        if (currVal !== "N") {

            // Create the left child for the current node
            currNode.left = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.left);
        }

        // For the right child
        i++;
        if (i >= str.length) break;
        currVal = str[i];

        // If the right child is not null
        if (currVal !== "N") {

            // Create the right child for the current node
            currNode.right = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.right);
        }
        i++;
    }

    return root;
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let input_ar1 = readLine().split(' ');
        let root = buildTree(input_ar1);
        let obj = new Solution();
        if (obj.isBST(root)) {
            console.log("true");
        } else {
            console.log("false");
        }
    }
}
// } Driver Code Ends


// User function Template for javascript

/**
 * @param {Node} root
 * @returns {boolean}
 */

class Solution {
    // Function to check whether a Binary Tree is BST or not.
//Your Code Start
    inorderTraversal(root) {
        if (root === null) {
            return true;
        }

        if (!this.inorderTraversal(root.left)) {
            return false;
        }

        if (this.prev !== null && root.data <= this.prev.data) {
            return false;
        }

        this.prev = root;

        return this.inorderTraversal(root.right);
    }

    isBST(root) {
        this.prev = null;
        return this.inorderTraversal(root);
    }
    }
//Your Code End
